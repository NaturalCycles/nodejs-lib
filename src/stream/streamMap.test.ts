import { _range, AggregatedError } from '@naturalcycles/js-lib'
import { readableFrom, streamMap } from '../index'

test('streamMap simple', async () => {
  const data = _range(1, 4).map(n => ({ id: String(n) }))

  const readable = readableFrom(data)

  const items: any[] = []
  const res = await streamMap(readable, async item => {
    items.push(item)
  })
  expect(res).toEqual([])
  expect(items).toEqual(data)
})

test('streamMap collectResults=true', async () => {
  const data = _range(1, 4).map(n => ({ id: String(n) }))

  const readable = readableFrom(data)

  const res = await streamMap(readable, async item => item, { collectResults: true })
  expect(res).toEqual(data)
  expect(readable.destroyed).toBe(true)
})

test('streamMap exception', async () => {
  const data = _range(1, 5).map(n => ({ id: String(n) }))

  const readable = readableFrom(data)

  await expect(
    streamMap(readable, async item => {
      if (item.id === '3') throw new Error('my error')
      return item
    }),
  ).rejects.toThrow('my error')

  expect(readable.destroyed).toBe(true)
})

test('streamMap exception stopOnError=false', async () => {
  const data = _range(1, 5).map(n => ({ id: String(n) }))

  const readable = readableFrom(data)

  const err: AggregatedError = await streamMap(
    readable,
    async item => {
      if (item.id === '3') throw new Error('my error')
      return item
    },
    { stopOnError: false, collectResults: true },
  ).catch(err => err)

  // console.log(err)
  expect(err).toMatchSnapshot()
  expect(err.errors).toMatchSnapshot()
  expect(err.results).toEqual(data.filter(r => r.id !== '3'))
  expect(readable.destroyed).toBe(true)
})

test('streamMap exception skipErrors=true', async () => {
  const data = _range(1, 5).map(n => ({ id: String(n) }))

  const readable = readableFrom(data)

  expect(
    await streamMap(
      readable,
      async item => {
        if (item.id === '3') throw new Error('my error')
        return item
      },
      { skipErrors: true, collectResults: true },
    ),
  ).toEqual(data.filter(r => r.id !== '3'))

  expect(readable.destroyed).toBe(true)
})
