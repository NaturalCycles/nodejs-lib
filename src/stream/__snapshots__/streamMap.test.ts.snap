// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`streamMap exception stopOnError=false 1`] = `
[AggregatedError: 1 errors:
1. my error]
`;

exports[`streamMap exception stopOnError=false 2`] = `
Array [
  [Error: my error],
]
`;
